vmm/src/vm.rs

```diff
@@ -883,25 +886,36 @@ impl Vm {
         Ok(vm)
     }
 
-    fn load_initramfs(&mut self, guest_mem: &GuestMemoryMmap) -> Result<arch::InitramfsConfig> {
-        let initramfs = self.initramfs.as_mut().unwrap();
-        let size: usize = initramfs
+    // Load initramfs or kernel image into memory
+    // @is_top: For initramfs is true as it always locates at the top memory address,
+    // for kernel, is true if there is no initramfs, otherwise false.
+    fn load_boot_file(
+        file: &mut Option<File>,
+        guest_mem: &GuestMemoryMmap,
+        is_top: bool,
+    ) -> Result<arch::BootFileConfig> {
+        let file = file.as_mut().unwrap();
+        let size: usize = file
             .seek(SeekFrom::End(0))
             .map_err(|_| Error::InitramfsLoad)?
             .try_into()
             .unwrap();
-        initramfs.rewind().map_err(|_| Error::InitramfsLoad)?;
 
-        let address =
+        let top_addr =
             arch::initramfs_load_addr(guest_mem, size).map_err(|_| Error::InitramfsLoad)?;
-        let address = GuestAddress(address);
+        #[cfg(target_arch = "aarch64")]
+        let offset = arch::aarch64::layout::KERNEL_IMAGE_MAX_SIZE;
+        #[cfg(target_arch = "x86_64")]
+        let offset = 0;
+        let address = if is_top { top_addr } else { top_addr - offset };
```

Comment:
This logic is only for AArch64, X86_64 has nothing with `top_addr` and `is_top`. Is it better if you calculate the address on the caller side and keep the logic of this function simple? Specifically:

1. Replace the parameter `is_top: bool,` with `addr: u64`
2. In this function, simply load the file to the specified `addr`
3. On X86_64, pass `arch::initramfs_load_addr()` when calling this function
4. On AArch64, apply your logic to determine the address in `configure_system()` before calling this function